rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.create/21_synthetic_source_stored.yml

```diff
@@ -353,11 +353,306 @@ stored field under object with store_array_source:
   - match: { hits.hits.1._source.obj: null }
 
 
+---
+field param - keep root array:
+  - requires:
+      cluster_features: ["mapper.synthetic_source_keep"]
+      reason: requires keeping array source
+
+  - do:
+      indices.create:
+        index: test
+        body:
+          mappings:
+            _source:
+              mode: synthetic
+            properties:
+              id:
+                type: integer
+              kw_default:
+                type: keyword
+              kw_arrays:
+                type: keyword
+                synthetic_source_keep: arrays
```

Comment:
What happens if `_source.mode` is `stored` and we use the `synthetic_source_keep` parameter? I expect that setting to be ignored completely. Mappings and settings are composed in templates so it is not unlikely that we will have mixed situations.

Reply:
What's the setting set to `stored`? I think I didn't get that part..

Reply:
This thing is called `synthetic_source_keep` but you can have an index that is not using synthetic source, right? In that case you would have a mapping parameter that is about synthetic source....but with no synthetic source (stored source). This might happen because of template and component template composition and to me looks a bit odd.

Also, some users might adopt an index mode (like logsdb or tsdb) which uses syntehtic source without actually knowing that synthetic source is used under the hood. Then you have to explain them that they need a parameter called `synthetic_source_keep` to preserve arrays...which requires telling them that synthetic source is used under the hood.

I don't know TBH what is the right thing her...it just looks a bit odd to me.

Reply:
Another thing is the distinction between arrays and singletons...users might not necessarily know if a field is a single-value field or a multi-value field because they not necessarily control the mapping. Think about fields coming from OTel or some integration...we are asking them to make a choice...and most of the time, just to be on the safe side, they will need to chose `all` because they don't know if the filed is single-value or multi-value.

Reply:
IMO we should have just two choices, `preserve_array`, `true` or `false`. I understand that not making a distinction between singleton and arrays might result in performance penalty...

Reply:
We had a long thread on this topic in https://github.com/elastic/elasticsearch/pull/112397.. Names are hard, we can argue for and against each option.

`all` should be used very deliberately, and probably rarely. The options for `arrays` is more widely applicable and easier to grasp, since this is a known limitation of synthetic source. It's thus better to have separate options for arrays and [arrays + singletons].